numpyro/handlers.py

```diff
@@ -184,7 +184,7 @@ def __init__(self, fn, guide_trace):
         super(replay, self).__init__(fn)
 
     def process_message(self, msg):
-        if msg['name'] in self.guide_trace:
+        if msg['name'] in self.guide_trace and msg['type'] in ('sample', 'plate'):
```

Comment:
Will this work if both `model` and `guide` have the same `params` site? I think it will work if we change the logic at [these lines](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/elbo.py#L60-L63) (basically removing `model_param_map` there). I think the performance would be the same but you certainly have a better idea than me regarding this. WDYT?

Reply:
Hmm..the only reason why I added this is because many of our effect handlers should implicitly operate on a subset of site types but we haven't specified that. Now that we have a few different types, I thought it will be good to specify that. `replay` shouldn't apply to param sites, but `substitute` is more general and allows for that. Let me know if I'm missing some use case you had in mind.

Reply:
At the logic in my previous comment, we filter out `params` available in `guide` and substitute to `model`. If we only replay `sample` sites here, for those params available in `guide`, we'll neither substitute or replay. I think if we add those checks, then we would need to `substitute` all params (including those params available in `guide`).

Reply:
I see. I just removed `model_param_map`, though please correct me if that's an oversimplification and I overlooked something. 😄

Reply:
I think it is correct. :)